sdks/go/pkg/beam/io/fileio/read.go

```diff
@@ -0,0 +1,126 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fileio
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/apache/beam/sdks/v2/go/pkg/beam"
+	"github.com/apache/beam/sdks/v2/go/pkg/beam/register"
+)
+
+func init() {
+	register.DoFn2x1[FileMetadata, func(ReadableFile), error](&readFn{})
+	register.Emitter1[ReadableFile]()
+}
+
+// directoryTreatment controls how paths to directories are treated when reading matches.
+type directoryTreatment int
+
+const (
+	// directorySkip skips directories.
+	directorySkip directoryTreatment = iota
+	// directoryDisallow disallows directories.
+	directoryDisallow
+)
+
+type readOption struct {
+	Compression        compressionType
+	DirectoryTreatment directoryTreatment
+}
+
+// ReadOptionFn is a function that can be passed to ReadMatches to configure options for
+// reading files.
+type ReadOptionFn func(*readOption)
+
+// ReadCompressionGzip specifies that files have been compressed using gzip.
+func ReadCompressionGzip() ReadOptionFn {
+	return func(o *readOption) {
+		o.Compression = compressionGzip
+	}
+}
+
+// ReadCompressionUncompressed specifies that files have not been compressed.
+func ReadCompressionUncompressed() ReadOptionFn {
```

Comment:
I think this still looks a bit ugly but I can't come up with a good short name. Only `ReadNoCompression` or `ReadWithoutCompression` which potentially read better, but then the functions would no longer have the shared prefix...

Reply:
I think it's fine to not overdo the "matching prefix" thing. 

`ReadGzip`
`ReadUncompressed` are pretty clear in context for users who want those options, and in the documentation make it clear that they are compression options, and that only one can be set at a time. One thing that none of the various option function consolidations do well is making mutual exclusivity clear in it's usage pattern vs in it's documentation. An option struct has it's own benefits and issues there but can force users to have vestigial option structs...

But at this stage, we should keep the functional options for consistency. Some future version of the SDK or some wrapper could do something like the [Proto MarshalOptions](https://pkg.go.dev/google.golang.org/protobuf/proto#MarshalOptions) are called out from the Options structs. But that innovation is recent WRT the SDK design...